python/adjoint/objective.py

```diff
@@ -1,193 +1,230 @@
         '''
-        self.sim = sim
+        src_center_dtft = np.matmul(
+            np.exp(1j * 2 * np.pi * np.array([src.frequency])[:, np.newaxis] *
+                   np.arange(y.size) * dt), y) * dt / np.sqrt(2 * np.pi)
+        adj_src_phase = np.exp(1j * np.angle(src_center_dtft))
+
+        if self._frequencies.size == 1:
+            # Single frequency simulations. We need to drive it with a time profile.
+            scale = dV * iomega / fwd_dtft / adj_src_phase  # final scale factor
+        else:
+            # multi frequency simulations
+            scale = dV * iomega / adj_src_phase
+        return scale
+
+    def _create_time_profile(self, fwidth_frac=0.1):
+        """Creates a time domain waveform for normalizing the adjoint source(s).
+
+        For single frequency objective functions, we should generate a guassian pulse with a reasonable
+        bandwidth centered at said frequency.
+
+        TODO:
+        The user may specify a scalar valued objective function across multiple frequencies (e.g. MSE) in
+        which case we should check that all the frequencies fit in the specified bandwidth.
+        """
+        return mp.GaussianSource(
+            np.mean(self._frequencies),
+            fwidth=fwidth_frac * np.mean(self._frequencies),
+        )
+
+
+class EigenmodeCoefficient(ObjectiveQuantity):
+    def __init__(self, sim, volume, mode, forward=True, **kwargs):
+        super().__init__(sim)
         self.volume = volume
         self.mode = mode
-        self.forward = 0 if forward else 1
-        self.normal_direction = None
-        self.kpoint_func = kpoint_func
-        self.eval = None
-        self.EigenMode_kwargs = kwargs
-        return
+        self.forward = forward
+        self.eigenmode_kwargs = kwargs
+        self._monitor = None
+        self._direction_vector = None
+        self._cscale = None
 
     def register_monitors(self, frequencies):
-        self.frequencies = np.asarray(frequencies)
-        self.monitor = self.sim.add_mode_monitor(frequencies,
-                                                 mp.ModeRegion(
-                                                     center=self.volume.center,
-                                                     size=self.volume.size),
-                                                 yee_grid=True)
-        self.normal_direction = self.monitor.normal_direction
-        return self.monitor
+        self._frequencies = np.asarray(frequencies)
+        self._monitor = self.sim.add_mode_monitor(
+            frequencies,
+            mp.ModeRegion(center=self.volume.center, size=self.volume.size),
+            yee_grid=True,
+        )
+        if self._monitor.normal_direction == 0:
+            direction_vector = mp.Vector3(x=1)
+        elif self._monitor.normal_direction == 1:
+            direction_vector = mp.Vector3(y=1)
+        elif self._monitor.normal_direction == 2:
+            direction_vector = mp.Vector3(z=1)
```

Comment:
Does this preclude us from passing arbitrary `kpoint`s in for oblique waveguides?

Reply:
I've reintroduced the `kpoint_func` param. WDYT?

Reply:
Seems right from a high level -- thanks for looking into for me.

Have you created a test for it by chance? Or are all the tests just forward/backward (or equally up/down)?

Reply:
The test case updated in this PR checks forward and backward gradients via the `forward` parameter of `EigenmodeCoefficient`, but does not have a case where `kpoint_func` is specified. I've personally not used the oblique waveguide capability but if you have something in mind let me know.